### PR TITLE
Implement AWS::RDS::DBSecurityGroup resource fixes #47

### LINKED ISSUE #47: ]

### PR DESCRIPTION
**Pull request by osdrv**
This PR provides a complete implementation for AWS::RDS::DBSecurityGroup
resource handler.

Resource specification:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-rds-security-group.html

Signed-off-by: Oleg Sidorov <sidorovo@amazon.com>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+        final Collection<Ingres> currentIngresses = dbSecurityGroup.get()
+                .ec2SecurityGroups()
+                .stream()
+                .map(Translator::translateDBSecurityGroupIngresFromSdk)
+                .collect(Collectors.toSet());
+
+        return progress
+                .then(p -> revokeIngresses(proxy, proxyClient, p, currentIngresses))
+                .then(p -> authorizeIngresses(proxy, proxyClient, p, desiredIngresses));
```
Comment (osdrv) [inline aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java:90]: This is suboptimal: we should find diff before triggering this update as it is done in the tag handler

**Review by moataz-mhmd — APPROVED** [review]

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            final CallbackContext callbackContext,
+            final ProxyClient<RdsClient> proxyClient,
+            final Logger logger) {
+
+        return ProgressEvent.progress(request.getDesiredResourceState(), callbackContext)
+                .then(progress -> {
+                    final ResourceModel model = progress.getResourceModel();
+                    if (StringUtils.isNullOrEmpty(model.getGroupName())) {
+                        model.setGroupName(
+                                IdentifierUtils.generateResourceIdentifier(
```
Comment (sungkkim) [inline aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java:32]: I would check that the format of this generated name matches existing resources. We've found issues in the past where customers relied on presence of stack-id in the primary identifier.

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                                        proxyClient.client()::createDBSecurityGroup
+                                );
+                            })
+                            .stabilize((createRequest, createResponse, proxyInvocation, model, context) -> {
+                                return isDBSecurityGroupStabilized(model, proxyInvocation);
+                            })
+                            .handleError((createRequest, exception, client, model, context) -> {
+                                if (exception instanceof DbSecurityGroupAlreadyExistsException) {
+                                    return ProgressEvent.defaultFailureHandler(exception, HandlerErrorCode.AlreadyExists);
+                                }
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java:64]: That's an interesting one: we ignore it in the existing impl. Would it make more sense to proxy to an update handler or ignore it completely? Maybe you can share a hint on when this situation can occur?

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                            .collect(Collectors.toSet());
+
+                    final Collection<Ingres> ingressToAdd = new ArrayList<>(desiredIngresses);
+                    final Collection<Ingres> ingressToRemove = new ArrayList<>(currentIngresses);
+
+                    ingressToRemove.removeAll(desiredIngresses);
+                    ingressToAdd.removeAll(currentIngresses);
+
+                    return ProgressEvent.progress(model, context)
+                            .then(p -> revokeIngresses(proxy, proxyClient, p, ingressToRemove))
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java:100]: I believe this is what's happening in `revokeIngresses/4`: it deals with a collection of items and ensures a stabilization between the items. Does it make sense?

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java]: Fixed. Thank you!

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                })
+                .handleError((request, exception, client, model, context) -> {
+                    if (exception instanceof DbSecurityGroupNotFoundException) {
+                        return ProgressEvent.defaultFailureHandler(exception, HandlerErrorCode.NotFound);
+                    }
+                    throw exception;
+                })
+                .done((request, response, proxyInvocation, model, context) -> {
+                    final DBSecurityGroup dbSecurityGroup = response.dbSecurityGroups().stream().findFirst().get();
+                    final Collection<Tag> existingTags = listTags(proxyClient, dbSecurityGroup.dbSecurityGroupArn());
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java]: Fixed. Thank you!

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java]: Fixed anyways.

File: aws-rds-dbsecuritygroup/aws-rds-dbsecuritygroup.json [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      "type": "array",
+      "uniqueItems": true,
+      "description": "Ingress rules to be applied to the DB security group.",
+      "items": {
+        "$ref": "#/definitions/Ingres"
+      }
+    },
+    "Tags": {
+      "type": "array",
+      "uniqueItems": true,
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/aws-rds-dbsecuritygroup.json]: Fixed. Thank you!

File: aws-rds-dbsecuritygroup/aws-rds-dbsecuritygroup.json [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        "rds:ModifyDBSecurityGroup",
+        "rds:RemoveTagsFromResource",
+        "rds:RevokeDBSecurityGroupIngress"
+      ]
+    },
+    "delete": {
+      "permissions": [
+        "rds:DeleteDBSecurityGroup",
+        "rds:DescribeDBSecurityGroups",
+        "rds:ListTagsForResource",
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/aws-rds-dbsecuritygroup.json]: Fixed, I assume. I'd appreciate a second look.

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/BaseHandlerStd.java:100]: Considering fixed.

File: aws-rds-dbsecuritygroup/aws-rds-dbsecuritygroup.json [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+          "minLength": 1,
+          "maxLength": 255
+        }
+      },
+      "required": [
+        "Key",
+        "Value"
+      ]
+    },
+    "Ingres": {
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/aws-rds-dbsecuritygroup.json]: Good catch and a pretty massive change. Fixed. Thank you!

File: aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+                                throw exception;
+                            })
+                            .progress();
+                })
+                .then(progress -> authorizeIngresses(proxy, proxyClient, progress, progress.getResourceModel().getDBSecurityGroupIngress()))
+                .then(progress -> new ReadHandler().handleRequest(proxy, request, callbackContext, proxyClient, logger));
```
[the remark this answers is not in the document]
Comment (osdrv) [reply aws-rds-dbsecuritygroup/src/main/java/software/amazon/rds/dbsecuritygroup/CreateHandler.java:62]: I had an impression that an account working with a resource via a CFN is expected to have a union set of IAM policies, not a subset. This made sense in the light of the legacy tagging implementation but I'm still having a hard time understanding this point. Could you please elaborate on it?

**Pull request closed without merging**